rep-2004.rst

```diff
@@ -0,0 +1,676 @@
+
+There's obviously a lot of ambiguity in this area, as you could argue for or against a variety of dependencies and how they affect the quality of a package.
+The goal is for the maintainers of a package to "make the case" that their dependencies are at least as high quality as the package in question.
+They should examine each dependency, and document any important caveats or justified exceptions in the package's "quality declaration" document so peer reviewers and consumers of the package can make their own evaluations.
+
+Quality Level 2
+^^^^^^^^^^^^^^^
+
+These are packages which need to be solidly developed and might be used in production environments, but are not strictly required, or are commonly replaced by custom solutions.
+This can also include packages which are not yet up to 'Level 1' but intend to be in the future.
+
+Package Requirements
+~~~~~~~~~~~~~~~~~~~~
+
+Requirements to be considered a 'Level 2' package:
+
+#. **Version Policy:**
+
+   #. The same as 'Level 1' packages
+
+#. **Change Control Process:**
+
+   #. Must have all code changes occur through a change request (e.g. pull request, merge request, etc.)
+   #. Must have confirmation of contributor origin (e.g. `DCO  <https://developercertificate.org/>`_, CLA, etc.)
+   #. Must have Continuous Integration (CI) policy for all change requests
+
+#. **Documentation:**
+
+   #. Must have documentation for each "feature" (e.g. for ``rclcpp``: create a node, publish a message, spin, etc.)
+   #. Must have a declared license or set of licenses
+   #. Must have a copyright statement in each source file
+   #. Must have a "quality declaration" document, which declares the quality level and justifies how the package meets each of the requirements
+
+      #. *Must have a section in the repository's ``README`` which contains the "quality declaration" or links to it*
+      #. *Must register with a centralized list of 'Level 2' packages, if one exists, to allow for peer review of the claim*
+
+#. **Testing:**
+
+   #. Must have system tests which cover all items in the "feature" documentation
+   #. Code Coverage:
+
+      #. *Must have code coverage tracking for the package*
+
+   #. Linters and Static Analysis
+
+      #. *Must have a code style and enforce it*
+      #. *Must use static analysis tools where applicable*
+
+#. **Dependencies:**
+
+   #. Must not have direct runtime "ROS" dependencies which are not 'Level 2' dependencies, but...
+   #. May have optional direct runtime "ROS" dependencies which are not 'Level 2', e.g. tracing or debugging features that can be disabled
+   #. Must have justification for why each direct runtime "non-ROS" dependency is equivalent to a 'Level 2' package in terms of quality
+
+#. **Platform Support:**
+
+   #. Must support all tier 1 platforms for ROS 2, as defined in `REP-2000 <https://www.ros.org/reps/rep-2000.html#support-tiers>`_
```

Comment:
Is there another way to word this (and the similar reqs below) in a way that includes ROS 1?

Reply:
How about:

> - Must support all target platforms for the package's ecosystem.
>   - For ROS 2 this means supporting all tier 1 platforms, as defined in ...

---

> (and the similar reqs below)

Can you point out each that you think has this issue?

Even this case (platform support), was meant to be only a milestone, with the idea that ROS 1 packages could simply state something like this in its quality declaration:

> This package only support Ubuntu X, Y, Z, as these are the only required platforms for ROS 1 according to REP-0003 (https://www.ros.org/reps/rep-0003.html).

But I take the point that as written it seems biased towards ROS 2. For the other cases though, we similarly tried to generalize them, so that's why I'm asking you to call out the other cases specifically.

Reply:
> Can you point out each that you think has this issue?
I just mean the other instances of this same point.

Reply:
Ah, sorry, I thought you meant other (different) requirements with the same issue (bias towards ROS 2).